### PR TITLE
Replaces const ref with && for apply

### PR DESCRIPTION
**Pull request by marymcbreen**
Addresses https://github.com/pytorch/pytorch/issues/5011
Tested with python test/test_autograd.py

### CODE REVIEW

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -345,14 +345,15 @@ static variable_list call_function(FunctionTask& task) {
   if(!task.base->keep_graph) {
     fn.will_release_variables();
   }
-  auto outputs = fn(inputs);
+  variable_list local_inputs = inputs;
+  auto outputs = fn(std::move(inputs));
```
Comment (apaszke) [inline torch/csrc/autograd/engine.cpp]: This has the downside that we now have an extra copy in the critical path, which wasn't there before. We should really avoid it when there are no callbacks registered (which is the usual case)

File: torch/csrc/autograd/function.cpp [hunk 2]
```diff
@@ -47,7 +47,8 @@ variable_list Function::traced_apply(variable_list inputs) {
 
   // Finally apply this Function.
   state_lock.unlock();
-  variable_list outputs = apply(inputs);
+  variable_list copied_inputs = inputs;
+  variable_list outputs = apply(std::move(copied_inputs));
```
Comment (apaszke) [inline torch/csrc/autograd/function.cpp]: Why do we need to copy them?

Comment (marymcbreen) [reply torch/csrc/autograd/function.cpp]: inputs can be used later in the function.

**Comment by marymcbreen** [comment]
Sorry for the churn on this.

File: torch/csrc/autograd/engine.cpp [hunk 3]
```diff
@@ -345,14 +345,29 @@ static variable_list call_function(FunctionTask& task) {
   if(!task.base->keep_graph) {
     fn.will_release_variables();
   }
-  auto outputs = fn(inputs);
+
+  const auto has_post_hooks = !fn.post_hooks().empty();
+  variable_list outputs;
+  if(has_post_hooks)
```
Comment (zdevito) [inline torch/csrc/autograd/engine.cpp]: nit:
```
if (has_post_hooks) {
 ...
} else {
 ...
}
```

File: torch/csrc/autograd/engine.cpp [hunk 4]
```diff
@@ -345,14 +345,29 @@ static variable_list call_function(FunctionTask& task) {
...
+    outputs = fn(std::move(inputs));
+  }
   validate_outputs(fn.next_edges(), outputs, [&](const std::string& msg) {
     std::ostringstream ss;
     ss << "Function "  << fn.name() << " returned an " << msg;
     return ss.str();
   });
   checkpoint_valid = prev_checkpoint_valid_state;
-  return call_post_hooks(fn, std::move(outputs), std::move(inputs));
+  if(has_post_hooks)
```
Comment (zdevito) [inline torch/csrc/autograd/engine.cpp]: nit: same formatting thing

**Review by zdevito — APPROVED** [review]
This looks good! I spot checked the places where move was added and didn't see any cases where the inputs were used after the move. A few minor formatting things and it is good to go.

**Comment by marymcbreen** [comment]
Updated for formatting.

**Pull request closed without merging**